server/scripts/dbNode.ts

```diff
@@ -392,29 +392,116 @@ async function mainProcess(producer: RabbitMqHandler, pool: Pool) {
   }
 }
 
-// begin process
-log.info('db-node initialization');
+let pool, producer, numWorkers, workerNumber;
+async function initializer(): Promise<void> {
+  // begin process
+  log.info('db-node initialization');
+
+  // setup sql client pool
+  pool = new Pool({
+    connectionString: DATABASE_URI,
+    ssl: {
+      rejectUnauthorized: false,
+    },
+    max: 3,
+  });
+
+  pool.on('error', (err, client) => {
+    log.error('Unexpected error on idle client', err);
+  });
+
+  // get all dyno's list
+  const res = await fetch(
+    'https://api.heroku.com/apps/commonwealth-staging2/dynos',
+    {
+      method: 'GET',
+      headers: {
+        Authorization: `Bearer ${process.env.HEROKU_API_TOKEN}`,
+        Accept: 'application/vnd.heroku+json; version=3',
+      },
+    }
+  );
+
+  if (!res.ok) {
+    log.info(`${res.status}, ${res.statusText}`);
+    throw new Error('Could not get dynoList');
+  }
+
+  const dynoList = await res.json();
+
+  if (!dynoList || dynoList.length === 0) {
+    // TODO: this will never occur
+    throw new Error("No dyno's detected");
+  }
+
+  // removes any dyno's that aren't ceNodes
+  const ceNodes = dynoList.filter((dyno) => dyno.name.includes('ceNode'));
 
-const producer = new RabbitMqHandler(RabbitMQConfig);
-const pool = new Pool({
-  connectionString: DATABASE_URI,
-  ssl: {
-    rejectUnauthorized: false,
-  },
-  max: 3,
-});
+  // sort CeNode dyno's by their id
+  ceNodes.sort((first, second) => {
+    if (first.id > second.id) return 1;
+    else if (first.id < second.id) return -1;
+    return 0;
+  });
 
-pool.on('error', (err, client) => {
-  log.error('Unexpected error on idle client', err);
-});
+  workerNumber = ceNodes
+    .map((dyno) => dyno.id)
+    .indexOf(process.env.HEROKU_DYNO_ID);
+  numWorkers = ceNodes.length;
+
+  let mostRecentDate = new Date(ceNodes[0].created_at);
+  let newestDyno = ceNodes[0];
+  for (const dyno of ceNodes) {
+    const dynoCreated = new Date(dyno.created_at);
+    if (mostRecentDate > dynoCreated) {
+      mostRecentDate = dynoCreated;
+      newestDyno = dyno;
+    }
+  }
 
-producer
-  .init()
+  if (
+    numWorkers !== Number(process.env.NUM_WORKERS) &&
+    newestDyno.id === process.env.HEROKU_DYNO_ID // prevents race condition by only allowing the most recently created dyno to update the config vars
+  ) {
+    const result = await fetch(
+      'https://api.heroku.com/apps/commonwealth-staging2/config-vars',
+      {
+        method: 'PATCH',
```

Comment:
Why do we even need the NUM_WORKERS environment variable if each dyno is already fetching that information via HTTP on dyno startup? We can use the numWorkers variable from above instead.

Reply:
To avoid an overlap scenario. So say for example I made this fetch request everytime a ceNode polls the database. Then every 10 minutes it would detect the new ceNodes that were added using the slider but during those 10 minutes the older ceNode would not be aware that a new ceNode was added thus the old ceNode is on the old numWorkers while the new ceNode has an updated numWorkers. This creates overlap where ceNodes will be listening to the same chains. To avoid this, whenever a new ceNode is created it updates the NUM_WORKERS which triggers a restart of all the other nodes, making them aware of the new node and avoiding overlap.

Reply:
It isn't necessary to use NUM_WORKERS if we don't mind the overlap and can filter duplicate events effectively.